onnxruntime/core/providers/cuda/tensor/slice.cc

```diff
@@ -119,22 +116,80 @@ Status Slice<dynamic>::ComputeInternal(OpKernelContext* ctx) const {
+  return Status::OK();
+}
+
+Status Impl(const void* input_data,
+            const TensorShape& input_shape,
+            void* output_data,
+            SliceOp::PrepareForComputeMetadata& compute_metadata,
+            size_t element_size) {
+  const auto& input_dimensions = input_shape.GetDims();
+  size_t dimension_count = input_dimensions.size();
+
+  TArray<int64_t> starts_buffer(compute_metadata.starts_);
+  TArray<int64_t> steps_buffer(compute_metadata.steps_);
+  TArray<int64_t> input_strides;
+  TArray<fast_divmod> output_strides;
+
+  ORT_RETURN_IF_ERROR(ComputeSliceStrides(input_shape, input_strides, output_strides, compute_metadata));
+
+  TensorShape output_shape(compute_metadata.output_dims_);
 
-  ORT_RETURN_IF_ERROR(CallSliceImp(element_size,
+  ORT_RETURN_IF_ERROR(SliceImpCore(input_data,
+                                   output_data,
+                                   element_size,
                                    gsl::narrow_cast<int32_t>(dimension_count),
                                    starts_buffer,
                                    steps_buffer,
                                    input_strides,
                                    output_strides,
-                                   ctx,
-                                   TensorShape(output_dims)));
+                                   output_shape));
+
+  return Status::OK();
+}
+}  // namespace SliceCuda
+
+template <bool dynamic>
+Status Slice<dynamic>::ComputeInternal(OpKernelContext* ctx) const {
+  const Tensor* input_tensor = GetSlicedOrUnslicedTensor(ctx);
+  ORT_ENFORCE(nullptr != input_tensor);
+  const auto& input_shape = input_tensor->Shape();
+  const auto& input_dimensions = input_shape.GetDims();
+  if (input_dimensions.empty()) return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT, "Cannot slice scalars");
+
+  SliceOp::PrepareForComputeMetadata compute_metadata(input_dimensions);
+
+  if (dynamic) {
+    std::vector<int64_t> input_starts, input_ends, input_axes, input_steps;
+    FillVectorsFromInput(*ctx->Input<Tensor>(1), *ctx->Input<Tensor>(2), ctx->Input<Tensor>(3),
+                         ctx->Input<Tensor>(4), input_starts, input_ends, input_axes, input_steps);
+    ORT_RETURN_IF_ERROR(PrepareForCompute(input_starts, input_ends, input_axes, input_steps, compute_metadata));
+
+  } else {
+    ORT_RETURN_IF_ERROR(PrepareForCompute(StartsAttribute(), EndsAttribute(), AxesAttribute(), compute_metadata));
+  }
+
+  TensorShape output_shape(compute_metadata.output_dims_);
+
+  TArray<int64_t> starts_buffer(compute_metadata.starts_);
+  TArray<int64_t> steps_buffer(compute_metadata.steps_);
+  TArray<int64_t> input_strides;
+  TArray<fast_divmod> output_strides;
+
+  ORT_RETURN_IF_ERROR(SliceCuda::ComputeSliceStrides(input_shape, input_strides, output_strides, compute_metadata));
+
+  ORT_RETURN_IF_ERROR(CallSliceImp(input_tensor->DataType()->Size(), input_dimensions.size(), starts_buffer,
```

Comment:
We can call `SliceCuda::Imp()` directly but we need to go through `CallSliceImp()` as `ComputeInternal()` is shared between the inferencing and training kernel and the training kernel overrides `CallSliceImp()`

Reply:
Would be good to have this as a comment in the code

---
In reply to: [462736008](https://github.com/microsoft/onnxruntime/pull/4627#discussion_r462736008) [](ancestors = 462736008)

Reply:
Done